src/lib.rs

```diff
@@ -126,8 +139,90 @@ pub fn select_coin_knapsack(
+    let mut accumulated_weight: u32 = 0;
+    let mut best_set: HashSet<(usize)> = HashSet::new();
+    // Assigning infinity to beginwith
+    let mut best_set_value: u64 = u64::MAX;
+    let mut best_set_weight: u32 = 0;
+    let mut rng = thread_rng();
+    for i in 1..=1000 {
+        for pass in 1..=2 {
+            if !target_reached {
+                for &(index, u) in smaller_coins {
+                    //Simulate a coin toss
+                    let toss_result: bool = rng.gen_bool(0.5);
+                    let selected_element = (index, u);
+                    if (pass == 2 && !selected_inputs.contains(&selected_element.0))
+                        || (pass == 1 && toss_result)
+                    {
+                        // Including the UTXO in the selected inputs
+                        selected_inputs.insert(selected_element.0);
+                        // Adding the effective value of an input (value - estimated fee)
+                        accumulated_value +=
+                            effective_value(&selected_element.1, options.target_feerate);
+                        accumulated_weight += u.weight;
+                        if accumulated_value == adjusted_target {
+                            // Perfect Match, Return the HashSet selected_inputs
+                            let estimated_fees =
+                                calculate_fee(accumulated_weight, options.target_feerate);
+                            let index_vector: Vec<usize> = selected_inputs.into_iter().collect();
+                            let waste: u64 = calculate_waste(
+                                inputs,
+                                &index_vector,
+                                &options,
+                                accumulated_value,
+                                accumulated_weight,
+                                estimated_fees,
+                            );
+                            return Ok(SelectionOutput {
+                                selected_inputs: index_vector,
+                                waste: WasteMetric(waste),
+                            });
+                        } else if accumulated_value >= adjusted_target {
+                            target_reached = true;
+                            if accumulated_value < best_set_value {
+                                // New best_set found
+                                best_set_value = accumulated_value;
+                                best_set_weight = accumulated_weight;
+                                best_set.clone_from(&selected_inputs);
+
+                                // Removing the last UTXO that raised selection_sum above adjusted_target to try to find a smaller set
+                                selected_inputs.remove(&selected_element.0);
+                                accumulated_value -= selected_element.1.value;
+                                accumulated_weight -= selected_element.1.weight;
```

Comment:
As you can see the `accumulated_weight` does not play any role in the selection of inputs in this algorithm.
So no need to waste CPU cycles here.